src/sema/expression.rs

```diff
@@ -1314,6 +1310,143 @@ fn coerce_number(
     })
 }
 
+fn number_literal(
+    loc: &pt::Loc,
+    base: &str,
+    exp: &str,
+    ns: &Namespace,
+    factor: &BigInt,
+    diagnostics: &mut Vec<Diagnostic>,
+    resolve_to: ResolveTo,
+) -> Result<Expression, ()> {
+    let base = BigInt::from_str(base).unwrap();
+
+    let n = if exp.is_empty() {
+        base
+    } else {
+        let b10 = BigInt::from_str("10").unwrap();
+
+        if let Some(exp) = exp.strip_prefix('-') {
+            if let Ok(exp) = u16::from_str(exp) {
+                let exp = b10.pow(exp);
+                let res = BigRational::new(base, exp);
+
+                if res.is_integer() {
+                    res.to_integer()
+                } else {
+                    return Ok(Expression::RationalNumberLiteral(*loc, Type::Rational, res));
+                }
+            } else {
+                diagnostics.push(Diagnostic::error(
+                    *loc,
+                    format!("exponent '{}' too large", exp),
+                ));
+                return Err(());
+            }
+        } else if let Ok(exp) = u16::from_str(exp) {
+            let exp = b10.pow(exp);
+
+            base.mul(exp)
```

Comment:
Are you checking if this number can fit in a `u256`?

Reply:
That's done elsewhere and there are test for that